zipline/algorithm.py

```diff
@@ -464,25 +464,38 @@ def run(self, source, overwrite_sim_params=True,
             # if DataFrame provided, map columns to sids and wrap
             # in DataFrameSource
             copy_frame = source.copy()
-            copy_frame.columns = \
-                self.asset_finder.map_identifier_index_to_sids(
-                    source.columns, source.index[0]
-                )
+            asset_sids = populate_finder_from_identifier_index(
+                self.asset_finder, source.columns, source.index[0])
+            copy_frame.columns = asset_sids
+            for sid in copy_frame.columns:
+                self.asset_finder.insert_metadata(sid)
+            self.asset_finder.populate_cache()
```

Comment:
Why is it necessary to both run pop_finder_from_ident_index as well as insert_metadata?